src/backend/gl/src/lib.rs

```diff
@@ -656,14 +656,31 @@ impl CommandQueue {
             },
             Command::CopyBuffer(src, dst, src_offset, dst_offset, size) => {
                 let gl = &self.share.context;
-                unsafe {
-                    gl.BindBuffer(gl::COPY_READ_BUFFER, src);
-                    gl.BindBuffer(gl::COPY_WRITE_BUFFER, dst);
-                    gl.CopyBufferSubData(gl::COPY_READ_BUFFER,
-                                         gl::COPY_WRITE_BUFFER,
-                                         src_offset,
-                                         dst_offset,
-                                         size);
+
+                if self.share.capabilities.copy_buffer_supported {
+                    unsafe {
+                        gl.BindBuffer(gl::COPY_READ_BUFFER, src);
+                        gl.BindBuffer(gl::COPY_WRITE_BUFFER, dst);
+                        gl.CopyBufferSubData(gl::COPY_READ_BUFFER,
+                                            gl::COPY_WRITE_BUFFER,
+                                            src_offset,
+                                            dst_offset,
+                                            size);
+                    }
+                } else {
+                    assert!(size > 0);
+
+                    unsafe {
+                        let mut src_ptr = 0 as *mut ::std::os::raw::c_void;
+                        factory::temporary_ensure_mapped(&mut src_ptr, gl::COPY_READ_BUFFER, src, memory::READ, gl);
```

Comment:
Are there any alternatives to glCopyBufferSubData than? Unless I'm misunderstanding, with a version of OpenGL that old, persistent mappings would not be supported and I wouldn't be able to upload data because the buffer is already mapped.